File: android/app/src/main/java/social/betty/features/browse/BrowseGrouping.kt
package social.betty.features.browse

import social.betty.core.model.PublicGroupItem

/** One rendered card on the Browse screen. */
sealed class BrowseCard {
    abstract val cardKey: String

    data class Single(val group: PublicGroupItem) : BrowseCard() {
        override val cardKey = "g-${group.id}"
    }

    data class TournamentBucket(
        val tournamentId: Int,
        val tournamentName: String,
        val tournamentImageUrl: String?,
        val groups: List<PublicGroupItem>,
    ) : BrowseCard() {
        override val cardKey = "t-$tournamentId"
    }
}

/**
 * Mirrors the web `visibleCards` bucketing logic (browse.vue) and BrowseGrouping.swift:
 *
 * - List mode → every item becomes a Single card.
 * - Grouped mode → items with a custom headerImageUrl stay Single (emitted first, in item
 *   order); the rest are bucketed by tournamentId in first-appearance order; a bucket of
 *   one collapses back to a Single.
 */
object BrowseGrouping {
    fun cards(items: List<PublicGroupItem>, grouped: Boolean): List<BrowseCard> {
        if (!grouped) return items.map { BrowseCard.Single(it) }

        val cards = mutableListOf<BrowseCard>()
        val bucketOrder = mutableListOf<Int>()
        val buckets = mutableMapOf<Int, MutableList<PublicGroupItem>>()

        for (item in items) {
            if (item.headerImageUrl != null) {
                cards += BrowseCard.Single(item)
                continue
            }
            if (!buckets.containsKey(item.tournamentId)) {
                bucketOrder += item.tournamentId
            }
            buckets.getOrPut(item.tournamentId) { mutableListOf() } += item
        }

        for (tId in bucketOrder) {
            val bucket = buckets[tId] ?: continue
            val first = bucket.firstOrNull() ?: continue
            if (bucket.size == 1) {
                cards += BrowseCard.Single(first)
            } else {
                cards += BrowseCard.TournamentBucket(
                    tournamentId = tId,
                    tournamentName = first.tournamentName ?: "",
                    tournamentImageUrl = first.tournamentImageUrl,
                    groups = bucket,
                )
            }
        }

        return cards
    }
}

File: ios/Betty/Features/GroupManagement/BrowseGrouping.swift
import Foundation

/// One card on the public-groups browse screen (web `BrowseCard` union).
nonisolated enum BrowseCard: Identifiable, Hashable, Sendable {
    case single(PublicGroupItem)
    case tournament(id: Int, name: String, imageURL: String?, groups: [PublicGroupItem])

    var id: String {
        switch self {
        case .single(let group): "g-\(group.id)"
        case .tournament(let id, _, _, _): "t-\(id)"
        }
    }
}

/// Web `visibleCards` bucketing (browse.vue, identical to the dashboard rule):
/// list mode → every item is a single card; grouped mode → items with a custom
/// `header_image_url` stay single (emitted first, in item order), the rest bucket by
/// `tournament_id` (first-appearance order); a bucket of one collapses back to a single.
nonisolated enum BrowseGrouping {
    static func cards(items: [PublicGroupItem], grouped: Bool) -> [BrowseCard] {
        guard grouped else {
            return items.map { .single($0) }
        }

        var cards: [BrowseCard] = []
        var bucketOrder: [Int] = []
        var buckets: [Int: [PublicGroupItem]] = [:]

        for item in items {
            if item.headerImageURL != nil {
                cards.append(.single(item))
                continue
            }
            if buckets[item.tournamentID] == nil {
                bucketOrder.append(item.tournamentID)
            }
            buckets[item.tournamentID, default: []].append(item)
        }

        for tournamentID in bucketOrder {
            guard let bucket = buckets[tournamentID], let first = bucket.first else { continue }
            if bucket.count == 1 {
                cards.append(.single(first))
            } else {
                cards.append(.tournament(
                    id: tournamentID,
                    name: first.tournamentName,
                    imageURL: first.tournamentImageURL,
                    groups: bucket
                ))
            }
        }

        return cards
    }
}
